File: question226/invert.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// had trouble with a case where you had to handle the null nodes... 
// 30m... 
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if (!root) return root;    

        stack<TreeNode*> l; 
        stack<TreeNode*> r; 
        TreeNode* lPtr = root; 
        TreeNode* rPtr = root; 

        while (!l.empty() || !r.empty() || lPtr || rPtr) {
            while (lPtr || rPtr) {
                l.push(lPtr); 
                r.push(rPtr); 

                lPtr = lPtr->left; 
                rPtr = rPtr->right; 
            }
            lPtr = l.top(); 
            rPtr = r.top(); 
            l.pop(); 
            r.pop(); 
            if (lPtr && rPtr) {
                int temp = lPtr->val; 
                lPtr->val = rPtr->val; 
                rPtr->val = temp; 
            } else if (lPtr && !rPtr) {
                
            } 
            
            lPtr = lPtr->right; 
            rPtr = rPtr->left; 

        }
        return root; 
    }
};

File: question226/invert.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# done in 19 minutes and its way easier to do in python as well as doing it in bfs than dfs 
class Solution(object):
    def invertTree(self, root):
        queue = [root]
        while (len(queue) != 0):
            if queue[0] is not None: 
                queue[0].left, queue[0].right = queue[0].right, queue[0].left
            
                if (queue[0].left is not None):
                    queue.append(queue[0].left)
                if (queue[0].right is not None): 
                    queue.append(queue[0].right)

            queue.pop(0)
        return root 